### PR TITLE
[Jenkins-47056] nested JNLPLauncher settings ignored

### PR DESCRIPTION
**Pull request by MartinNowak**
See [JENKINS-47056](https://issues.jenkins-ci.org/browse/JENKINS-47056).

fixed by stripping off DelegatedComputerLaunchers to render slave-agent.jnlp

### Proposed changelog entries

* Entry 1: fixed nested JNLPLauncher settings (e.g. slave-setup plugin)

### CODE REVIEW

File: core/src/main/java/hudson/slaves/SlaveComputer.java [hunk 1]
```diff
@@ -234,10 +234,25 @@ public boolean isLaunchSupported() {
...
     }
 
+    /**
+     * Return the {@code ComputerLauncher} for this SlaveComputer, strips off
+     * any {@code DelegatingComputerLauncher}s.
+     */
+    public ComputerLauncher getDelegatedLauncher() {
+        ComputerLauncher l = launcher;
+        while (l instanceof DelegatingComputerLauncher) {
+            l = ((DelegatingComputerLauncher)l).getLauncher();
```
Comment (oleg-nenashev) [inline core/src/main/java/hudson/slaves/SlaveComputer.java]: What if it the nested launcher is a delegating launcher as well?
I would expect the method to continue the recursion then.

File: core/src/main/java/hudson/slaves/SlaveComputer.java [hunk 2]
```diff
@@ -234,10 +234,25 @@ public boolean isLaunchSupported() {
...
+     * Return the {@code ComputerLauncher} for this SlaveComputer.
+     */
     public ComputerLauncher getLauncher() {
         return launcher;
     }
 
+    /**
+     * Return the {@code ComputerLauncher} for this SlaveComputer, strips off
+     * any {@code DelegatingComputerLauncher}s.
+     */
```
Comment (oleg-nenashev) [inline core/src/main/java/hudson/slaves/SlaveComputer.java]: `@since TODO`

File: test/src/test/java/hudson/slaves/JNLPLauncherTest.java [hunk 3]
```diff
@@ -212,23 +234,20 @@ private void launchJnlpAndVerify(Computer c, ArgumentListBuilder args) throws Ex
     }
 
     /**
-     * Determines the link to the .jnlp file.
+     * Adds a JNLP {@link Slave} to the system and returns it.
      */
-    private String getJnlpLink(Computer c) throws Exception {
```
Comment (oleg-nenashev) [inline test/src/test/java/hudson/slaves/JNLPLauncherTest.java]: why would you remove this method? Seems unrelated

**Review by oleg-nenashev — CHANGES_REQUESTED** [review]
The approach looks reasonable to me though it is a bit hackish. Multiple levels of launcher nesting should be addressed then

File: test/src/test/java/hudson/slaves/JNLPLauncherTest.java (hunk 3 above)
Comment (MartinNowak) [reply test/src/test/java/hudson/slaves/JNLPLauncherTest.java]: The view for a nested launcher is different (actually none for the unit-test launcher) and therefor it's not possible to scrape the JNLP link from the page. Also the scraping was hardly necessary, the link hasn't changed since 10 years, so I just replaced it with a hard-coded URL path.

File: core/src/main/java/hudson/slaves/SlaveComputer.java (hunk 1 above)
Comment (MartinNowak) [reply core/src/main/java/hudson/slaves/SlaveComputer.java]: That's why it's while loop ;).

File: core/src/main/java/hudson/slaves/SlaveComputer.java (hunk 2 above)
Comment (MartinNowak) [reply core/src/main/java/hudson/slaves/SlaveComputer.java]: Seems like the next version is 2.80, correct?

File: core/src/main/java/hudson/slaves/SlaveComputer.java (hunk 1 above)
Comment (MartinNowak) [reply core/src/main/java/hudson/slaves/SlaveComputer.java]: Also included ComputerLauncherFilter which has the same problem.

Comment (oleg-nenashev) [reply core/src/main/java/hudson/slaves/SlaveComputer.java]: Sorry, probably I was blind :(

File: test/src/test/java/hudson/slaves/JNLPLauncherTest.java (hunk 3 above)
Comment (oleg-nenashev) [reply test/src/test/java/hudson/slaves/JNLPLauncherTest.java]: agreed

File: core/src/main/java/hudson/slaves/SlaveComputer.java (hunk 2 above)
Comment (oleg-nenashev) [reply core/src/main/java/hudson/slaves/SlaveComputer.java]: Well, the train departs today. Hard to say if it gets into this release or not

**Review by oleg-nenashev — APPROVED** [review]
👍 excepting the hardcoded `@since`

File: core/src/main/java/hudson/slaves/SlaveComputer.java [hunk 4]
```diff
@@ -234,10 +234,33 @@ public boolean isLaunchSupported() {
...
+     * @since 1.312
+     */
     public ComputerLauncher getLauncher() {
         return launcher;
     }
 
+    /**
+     * Return the {@code ComputerLauncher} for this SlaveComputer, strips off
+     * any {@code DelegatingComputerLauncher}s or {@code ComputerLauncherFilter}s.
+     * @since 2.80
```
Comment (daniel-beck) [inline core/src/main/java/hudson/slaves/SlaveComputer.java]: I have bad news for you 😉

**Comment by MartinNowak** [comment]
Bumped the version to 2.81.
Guess with weakly releases you have quite some churn like that (https://github.com/jenkinsci/jenkins/search?q=%22%40since+TODO%22&type=Commits&utf8=%E2%9C%93).

How about standardizing `@since TODO` (mention it in PULL_REQUEST_TEMPLATE.md) and then running a simple
```sh
find -name '*.java' | xargs sed -i 's|@since TODO|2.81|g'
```
before tagging.

**Comment by oleg-nenashev** [comment]
yeah, it's something we need to do at some point.
Ideally it needs a Javadoc checker which would enforce it for new APIs in PR builders. But oh well, there are so many such low-hanging fruits

**Comment by oleg-nenashev** [comment]
Ready to go IMHO

**Comment by MartinNowak** [comment]
Bumped the version again to 2.83, if it changes again (hopefully not), then simply just edit it before merging.

**Comment by daniel-beck** [comment]
Was this branched off a bad master, or why are the #3007 test failures in here?

**Comment by MartinNowak** [comment]
No idea, the build was marked unstable not as failed. Rebased on current master, hopefully that fixes it.
